src/common/cockpitframe.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2014 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include "cockpitframe.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define MAX_FRAME_SIZE_BYTES 7
+
+/**
+ * cockpit_frame_parse:
+ * @input: An buffer of bytes
+ * @length: The length of @input buffer
+ * @consumed: Number of bytes consumed from @input
+ *
+ * Parse message framing length string from the top
+ * of the @input buffer. These are used by Cockpit transport framing
+ * over a stream based protocol.
+ *
+ * Returns: The length, zero if more data is needed, or -1 if an error.
+ */
+ssize_t
+cockpit_frame_parse (unsigned char *input,
+                     size_t length,
+                     size_t *consumed)
+{
+  size_t size = 0;
+  size_t i;
+
+  assert (input != NULL || length == 0);
+
+  size = 0;
+  for (i = 0; i < length; i++)
+    {
+      /* Check invalid characters, prevent integer overflow, limit max length */
+      if (i > MAX_FRAME_SIZE_BYTES || (char)(input[i]) < '0' || (char)(input[i]) > '9')
+        break;
+      size *= 10;
+      size += (char)(input[i]) - '0';
+    }
+
+  /* Want more data */
+  if (i == length)
+    return 0;
+
+  /* A failure */
+  if (size == 0 || input[i] != '\n')
+    return -1;
+
+  if (consumed)
+    *consumed = i + 1;
+  return size;
+}
+
+static ssize_t
+write_all (int fd,
+           unsigned char *data,
+           size_t length)
```

Comment:
Well it would need to be moved a file that doesn't use GLib, since the goal is to use the cockpitframe.c code in cockpit-session. But I can try and do that.